src/stats.c

```diff
@@ -64,13 +64,21 @@ extern inline Obj EXEC_CURR_FUNC(void)
     return result;
 }
 
+// The next macro is used in loop bodies to handle 'break', 'continue' and
+// 'return' statements. Note that here, during execution, `status` can only
+// take on the value STATUS_BREAK, STATUS_CONTINUE, STATUS_RETURN, and
+// STATUS_END; while STATUS_EOF, STATUS_ERROR, STATUS_QUIT, STATUS_QQUIT are
+// impossible.
+//
+// The checks are arranged so that the most frequent cases (i.e. a statement
+// which is not break/continue/return) is handled first.
 #define EXEC_STAT_IN_LOOP(stat)                                              \
     {                                                                        \
         UInt status = EXEC_STAT(stat);                                       \
-        if (status != 0) {                                                   \
+        if (status != STATUS_END) {                                          \
             if (status == STATUS_CONTINUE)                                   \
                 continue;                                                    \
-            return (status & (STATUS_RETURN_VAL | STATUS_RETURN_VOID));      \
+            return (status == STATUS_RETURN) ? STATUS_RETURN : STATUS_END;   \
```

Comment:
This part of the change is actually a bit subtle and I initially got it wrong -- and it still worked *by pure chance*. What I had done was this:
```suggestion
            return status == STATUS_RETURN
```
Which actually does the same as long as `STATUS_RETURN` is 1 and `STATUS_END` is 0, as is the case right now. But when I later experimented with changing these values, I got weird errors that I didn't understand at first.

I've now added a comment explaining what's going on here...